glean-core/src/histogram.rs

```diff
@@ -0,0 +1,222 @@
+///
+/// ```rust,ignore
+/// let mut hist = Histogram::exponential(1, 500, 10);
+///
+/// for i in 1..=10 {
+///     hist.add(i);
+/// }
+///
+/// assert_eq!(10, hist.count());
+/// assert_eq!(55, hist.sum());
+/// ```
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Histogram {
+    min: u32,
+    max: u32,
+    ranges: Vec<u32>,
+    buckets: Vec<u32>,
+
+    count: u32,
+    sum: u32,
+    typ: Type,
+}
+
+impl Default for Histogram {
+    fn default() -> Histogram {
+        Histogram::exponential(DEFAULT_RANGE_MIN, DEFAULT_RANGE_MAX, DEFAULT_BUCKET_COUNT)
+    }
+}
+
+impl Histogram {
+    /// Create a histogram with `count` exponential buckets in the range `min` to `max`.
+    pub fn exponential(min: u32, max: u32, count: u32) -> Histogram {
+        let ranges = exponential_range(min, max, count);
+
+        Histogram {
+            min,
+            max,
+            ranges,
+            buckets: vec![0; count as usize],
+            count: 0,
+            sum: 0,
+            typ: Type::Exponential,
+        }
+    }
+
+    /// Get the number of buckets in this histogram.
+    pub fn bucket_count(&self) -> usize {
+        self.buckets.len()
+    }
+
+    /// Add a single value to this histogram.
+    pub fn accumulate(&mut self, sample: u32) {
+        *self.bucket(sample) += 1;
+        self.sum += sample;
+        self.count += 1;
+    }
+
+    /// Get the total sum of values recorded in this histogram.
+    pub fn sum(&self) -> u32 {
+        self.sum
+    }
+
+    /// Get the total count of values recorded in this histogram.
+    pub fn count(&self) -> u32 {
+        self.count
+    }
+
+    fn bucket(&mut self, value: u32) -> &mut u32 {
+        let mut under = 0;
+        let mut over = self.bucket_count();
+        let mut mid;
+
+        loop {
+            mid = under + (over - under) / 2;
+            if mid == under {
+                break;
+            }
+            if self.ranges()[mid] <= value {
+                under = mid;
+            } else {
+                over = mid;
+            }
+        }
+
+        &mut self.buckets[mid]
+    }
+
+    /// Get the list of ranges.
+    pub fn ranges(&self) -> &[u32] {
+        &self.ranges
+    }
+
+    /// Get the filled buckets.
+    pub fn buckets(&self) -> &[u32] {
+        &self.buckets
+    }
+
+    /// Get the minimal recordable value.
+    pub fn min(&self) -> u32 {
+        self.min
+    }
+
+    /// Get the maximal recordable value.
+    pub fn max(&self) -> u32 {
+        self.max
+    }
+
+    /// Get this histogram's type.
+    pub fn typ(&self) -> Type {
+        self.typ
+    }
+
+    /// Clear the stored data in the histogram
+    pub fn clear(&mut self) {
+        self.sum = 0;
+        self.count = 0;
+
+        for bucket in self.buckets.iter_mut() {
+            *bucket = 0;
+        }
+    }
+
+    /// Check if this histogram recorded any values.
+    pub fn is_empty(&self) -> bool {
+        self.count() == 0
+    }
+}
+
+#[cfg(test)]
```

Comment:
Unit tests inline (everything that can be tested self-contained).
Integration tests in `tests/`.

`Histogram` is a crate-internal only thing. We can't test it from the outside (unless we actually export it from the library which I want to avoid)

Reply:
This totally makes sense. We should probably have a section about that in the developer docs.